src/index.rs

```diff
@@ -1,38 +1,141 @@
 use rio_api::parser::TriplesParser;
 use rio_turtle::TurtleError;
-use std::{io::Write, path::Path};
+use serde::{Deserialize, Serialize};
+use smallvec::{smallvec, SmallVec};
+use std::{
+    collections::HashMap,
+    hash::{DefaultHasher, Hash, Hasher},
+    path::Path,
+};
 
 use crate::{
     io,
     rdf_types::{Triple, TripleView},
 };
 
-fn index_triple(t: Triple, out: &mut impl Write) {
+/// Stores a mapping from hashed instance uri to their types
+#[derive(Serialize, Deserialize)]
+pub struct TypeIndex {
+    pub types: Vec<String>,
+    map: HashMap<u64, SmallVec<[usize; 1]>>,
+}
+
+impl TypeIndex {
+    fn hash(&mut self, s: &str) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        s.hash(&mut hasher);
+        hasher.finish().to_le()
+    }
+
+    pub fn from_iter(type_map: impl Iterator<Item = (String, String)>) -> Self {
+        let vec: Vec<(String, String)> = type_map.collect();
+        let mut idx = TypeIndex::new();
+        idx.types = vec
+            .clone()
+            .iter()
+            .map(|(_, t)| t.clone())
+            .collect::<std::collections::HashSet<String>>()
+            .into_iter()
+            .collect();
+
+        vec.iter().for_each(|(subject, type_uri)| {
+            idx.insert(&subject.to_string(), &type_uri.to_string())
+                .unwrap()
+        });
+
+        return idx;
+    }
+
+    pub fn new() -> Self {
+        TypeIndex {
+            types: Vec::new(),
+            map: HashMap::new(),
+        }
+    }
+
+    // Insert input subject-type mapping into the index.
+    // The index will store the hash of the subject.
+    pub fn insert(&mut self, subject_uri: &str, type_uri: &str) -> Result<(), std::io::Error> {
+        let key = self.hash(subject_uri);
+        let type_idx: usize;
+
+        // Get type index or add a new one
+        if self.types.contains(&type_uri.to_string()) {
```

Comment:
hint: This is getting O(m*n) with `m`-lookup and `n`-types: maybe the following is better:
```rust
pub struct TypeIndex {
    pub types: Vec<String>,
    type_indices: Option<HashMap<u64, usize>>,       // Mapping `hash(type_uri)` -> index in `types`
    map: HashMap<u64, SmallVec<[usize; 1]>>,// Mapping `hash(subject_uri)` -> indices in `types`
}
```

On `new` you would init `type_indices` and use it in `insert` to check the if the type is already existing. Provide an `optimize()` function on `TypeIndex` which will set the option to `None` to remove the unneeded space, also do not serialize it out to the YAML.

Reply:
Captured in #58  and keeping for a separate PR, as this will require additional refactoring